File: src/wasm/path_patching.cpp
#include "path_patching.h"
#include "raycasting.h"
#include "nav_utils.h"
#include "path_corridor.h"
#include "constants_layout.h"
#include "math_utils.h"
#include <algorithm>
 

extern Navmesh g_navmesh;

static inline void append_tris_as_polys(const std::vector<int>& tris, std::vector<int>& out) {
  for (int i = tris.size() - 1; i >= 0; --i) {
    const int p = g_navmesh.triangle_to_polygon[tris[i]];
    if (out.empty() || out.back() != p) {
      out.push_back(p);
    }
  }
}

// Efficient two-corridor merge: first corridor must begin at the join polygon,
// second continues from there back toward the agent. Writes result into outPolyCorr.
static inline bool merge_corridors(
  const std::vector<int>& triCorrFirst,
  const std::vector<int>& triCorrSecond,
  const std::vector<int>& originalCorr,
  int joinTriangle,
  std::vector<int>& outPolyCorr
) {
  outPolyCorr.clear();

  // Determine where to rejoin the original corridor (using polygon index of the join triangle)
  const int joinPoly = g_navmesh.triangle_to_polygon[joinTriangle];

  int originalCorridorJoinIndex = -1;
  for (int i = static_cast<int>(originalCorr.size()) - 1; i >= 0; --i) {
    if (originalCorr[i] == joinPoly) {
      originalCorridorJoinIndex = i;
      break;
    }
  }

  // Preserve the original corridor prefix up to (but excluding) the join polygon.
  if (originalCorridorJoinIndex != -1) {
    // Conservative reserve to reduce reallocations
    outPolyCorr.reserve(
      (originalCorridorJoinIndex > 0 ? originalCorridorJoinIndex : 0)
      + static_cast<int>(triCorrFirst.size())
      + static_cast<int>(triCorrSecond.size())
    );
    outPolyCorr.insert(outPolyCorr.end(), originalCorr.begin(), originalCorr.begin() + originalCorridorJoinIndex);
  }

  // Then append the new (triangle) corridors converted to polygon corridors in the given order.
  append_tris_as_polys(triCorrFirst, outPolyCorr);
  append_tris_as_polys(triCorrSecond, outPolyCorr);

  return !outPolyCorr.empty();
}

static inline int vertex_index_from_point_on_triangle(int triIdx, const Point2& p)
{
  const int v1 = g_navmesh.triangles[triIdx * 3 + 0];
  const int v2 = g_navmesh.triangles[triIdx * 3 + 1];
  const int v3 = g_navmesh.triangles[triIdx * 3 + 2];
  const Point2& P1 = g_navmesh.vertices[v1];
  const Point2& P2 = g_navmesh.vertices[v2];
  const Point2& P3 = g_navmesh.vertices[v3];
  const float eps = 1e-5f;
  auto close = [&](const Point2& a, const Point2& b){ return std::abs(a.x - b.x) < eps && std::abs(a.y - b.y) < eps; };
  if (close(p, P1)) return v1;
  if (close(p, P2)) return v2;
  if (close(p, P3)) return v3;
  return -1;
}

static inline bool compute_corner_miter_offset(int polyIdx, int cornerVIdx, const Point2& cornerPoint, float offset, Point2& outPoint)
{
  const int start = g_navmesh.polygons[polyIdx];
  const int end = g_navmesh.polygons[polyIdx + 1];
  for (int i = start; i < end; ++i) {
    if (g_navmesh.poly_verts[i] == cornerVIdx) {
      const int prevIndex = (i == start) ? end - 1 : i - 1;
      const int nextIndex = (i == end - 1) ? start : i + 1;
      const Point2 A = g_navmesh.vertices[g_navmesh.poly_verts[prevIndex]];
      const Point2 C = g_navmesh.vertices[g_navmesh.poly_verts[nextIndex]];

      Point2 vBA = cornerPoint - A;
      Point2 vBC = cornerPoint - C;
      math::normalize_inplace(vBA);
      math::normalize_inplace(vBC);
      Point2 miter = {vBA.x + vBC.x, vBA.y + vBC.y};
      if (math::length_sq(miter) <= 1e-12f) return false;
      math::normalize_inplace(miter);
      outPoint = {cornerPoint.x + miter.x * offset, cornerPoint.y + miter.y * offset};
      return true;
    }
  }
  return false;
}

bool attempt_path_patch(
  Navmesh& navmesh,
  int idx,
  int hitV1_idx,
  int hitV2_idx,
  int hitTri_idx,
  const std::vector<int>& raycastTriCorridor
) {
  if (raycastTriCorridor.empty()) return false;

  // Build hit edge points from vertex indices for geometric checks
  const Point2 hitP1 = g_navmesh.vertices[hitV1_idx];
  const Point2 hitP2 = g_navmesh.vertices[hitV2_idx];

  // Unwalkable triangle that blocked the ray
  const int blockingTri = hitTri_idx;
  const int blockingPoly = (blockingTri != -1) ? g_navmesh.triangle_to_polygon[blockingTri] : -1;

  // Approach 2: miter-offset around solid/obstacle polygon hit
  if (blockingPoly >= g_navmesh.walkable_polygon_count) {
    
    const float d1 = math::distancePointToSegment(hitP1, agent_data.last_visible_points_for_next_corner[idx], agent_data.next_corners[idx]);
    const float d2 = math::distancePointToSegment(hitP2, agent_data.last_visible_points_for_next_corner[idx], agent_data.next_corners[idx]);
    const bool useFirst = (d1 <= d2);
    const Point2 chosenCornerPoint = useFirst ? hitP1 : hitP2;
    const int chosenVIdx = useFirst ? hitV1_idx : hitV2_idx;

    if (chosenVIdx != -1) {
      Point2 offsetPoint;
      if (compute_corner_miter_offset(blockingPoly, chosenVIdx, chosenCornerPoint, CORNER_OFFSET, offsetPoint)) {
        const int offsetTri = getTriangleFromPoint(offsetPoint);
        if (offsetTri != -1) {
          RaycastCorridorResult rc1 = raycastCorridor(agent_data.positions[idx], offsetPoint, agent_data.current_tris[idx], offsetTri);
          if (rc1.hitV1_idx == -1 && !rc1.corridor.empty()) {
            // If we already have two corners, ensure we can reach nextCorner2
            if (agent_data.num_valid_corners[idx] == 2) {
              RaycastCorridorResult rc3 = raycastCorridor(offsetPoint, agent_data.next_corners2[idx], offsetTri, agent_data.next_corner_tris2[idx]);
              if (rc3.hitV1_idx != -1 || rc3.corridor.empty()) {
                // fallback to trying nextCorner first
                RaycastCorridorResult rc2 = raycastCorridor(offsetPoint, agent_data.next_corners[idx], offsetTri, agent_data.next_corner_tris[idx]);
                if (rc2.hitV1_idx != -1 || rc2.corridor.empty()) {
                  // give up miter approach
                } else {
                  // Update corners: insert offset as nextCorner, keep nextCorner2 as is
                  // Preserve the old nextCorner triangle for correct rejoin point
                  const int oldNextCornerTri = agent_data.next_corner_tris[idx];
                  agent_data.next_corners[idx] = offsetPoint;
                  agent_data.next_corner_tris[idx] = offsetTri;
                  std::vector<int> merged;
                  // IMPORTANT: rejoin at old nextCorner, not nextCorner2
                  if (merge_corridors(rc2.corridor, rc1.corridor, agent_data.corridors[idx], oldNextCornerTri, merged)) {
                    agent_data.corridors[idx] = std::move(merged);
                    return true;
                  }
                }
              } else {
                // Great: we can go offset -> nextCorner2; set nextCorner to offset and keep nextCorner2
                agent_data.next_corners[idx] = offsetPoint;
                agent_data.next_corner_tris[idx] = offsetTri;
                std::vector<int> merged;
                if (merge_corridors(rc3.corridor, rc1.corridor, agent_data.corridors[idx], agent_data.next_corner_tris2[idx], merged)) {
                  agent_data.corridors[idx] = std::move(merged);
                  return true;
                }
              }
            } else {
              // Only one corner known: must be able to go offset -> nextCorner
              RaycastCorridorResult rc2 = raycastCorridor(offsetPoint, agent_data.next_corners[idx], offsetTri, agent_data.next_corner_tris[idx]);
              if (rc2.hitV1_idx == -1 && !rc2.corridor.empty()) {
                agent_data.next_corners2[idx] = agent_data.next_corners[idx];
                agent_data.next_corner_tris2[idx] = agent_data.next_corner_tris[idx];
                agent_data.next_corners[idx] = offsetPoint;
                agent_data.next_corner_tris[idx] = offsetTri;
                agent_data.num_valid_corners[idx] = 2;

                std::vector<int> merged;
                if (merge_corridors(rc2.corridor, rc1.corridor, agent_data.corridors[idx], agent_data.next_corner_tris2[idx], merged)) {
                  agent_data.corridors[idx] = std::move(merged);
                  return true;
                }
              }
            }
          }
        }
      }
    }
  }

  // Approach 1: Intersection-based patch
  {
    
    const Point2& L = agent_data.last_visible_points_for_next_corner[idx];
    const Point2& C = agent_data.next_corners[idx];
    const Point2& A = agent_data.positions[idx];

    Point2 d1 = math::normalize(C - L);
    Point2 edgeDir = math::normalize(hitP2 - hitP1);
    const float dp = std::abs(math::dot(d1, edgeDir));
    if (dp <= 0.8f) {
      const float denom = math::cross(d1, edgeDir);
      if (std::abs(denom) > 1e-6f) {
        const Point2 LA = A - L;
        const float t = math::cross(LA, edgeDir) / denom;
        Point2 R = L + d1 * t;
        const float dR2 = math::distance_sq(A, R);
        const float dC2 = math::distance_sq(A, C);
        if (dR2 <= dC2 * 2.25f) {
          const int rTri = getTriangleFromPoint(R);
          if (rTri != -1) {
            RaycastCorridorResult rc1 = raycastCorridor(A, R, agent_data.current_tris[idx], rTri);
            RaycastCorridorResult rc2 = raycastCorridor(R, agent_data.next_corners[idx], rTri, agent_data.next_corner_tris[idx]);
            if (rc1.hitV1_idx == -1 && rc2.hitV1_idx == -1 && !rc1.corridor.empty() && !rc2.corridor.empty()) {
              agent_data.next_corners2[idx] = agent_data.next_corners[idx];
              agent_data.next_corner_tris2[idx] = agent_data.next_corner_tris[idx];
              agent_data.next_corners[idx] = R;
              agent_data.next_corner_tris[idx] = rTri;
              agent_data.num_valid_corners[idx] = 2;

              std::vector<int> merged;
              if (merge_corridors(rc2.corridor, rc1.corridor, agent_data.corridors[idx], agent_data.next_corner_tris2[idx], merged)) {
                agent_data.corridors[idx] = std::move(merged);
                return true;
              }
            }
          }
        }
      }
    }
  }

  return false;
} 

File: src/logic/agents/PathPatching.ts
import { copy, distance_sq, distancePointToSegment, dot, Line, lineLineIntersection, normalize_, Point2, set_, subtract_ } from "../core/math";
import { ACBLACK, ACBLUE, ACEMERALD, ACINDIGO, ACMAGENTA, ACPINK, ACRED, ACYELLOW, sceneState } from "../drawing/SceneState";
import { Navmesh } from "../navmesh/Navmesh";
import { getTriangleFromPoint } from "../navmesh/NavUtils";
import { offsetCorner } from "../navmesh/pathCorners";
import { raycastCorridor, RaycastWithCorridorResult } from "../Raycasting";
import { Agent } from "./Agent";
import { NavConst } from "./NavConst";

const line1Direction: Point2 = { x: 0, y: 0 };
const line1: Line = { point: { x: 0, y: 0 }, direction: { x: 0, y: 0 } };
const line2: Line = { point: { x: 0, y: 0 }, direction: { x: 0, y: 0 } };

/**
 * Calculates the intersection point between two lines
 * LINE1: from lastVisiblePoint to nextCorner
 * LINE2: from agentPosition parallel to blockingEdge
 * 
 * @param lastVisiblePoint - Last point where the next corner was visible
 * @param nextCorner - The target corner we're trying to reach
 * @param agentPosition - Current agent position
 * @param blockingEdgeDir - Direction vector of the blocking edge (normalized)
 * @returns Intersection point or null if lines are parallel/don't intersect properly
 */
export function calculatePathPatchIntersection(
  lastVisiblePoint: Point2,
  nextCorner: Point2,
  agentPosition: Point2,
  blockingEdgeDir: Point2
): Point2 | null {
  set_(line1Direction, nextCorner);
  subtract_(line1Direction, lastVisiblePoint);
  normalize_(line1Direction);
  
  const dotProduct = Math.abs(dot(line1Direction, blockingEdgeDir));
  if (dotProduct > 0.8) {
    return null;
  }

  set_(line1.point, lastVisiblePoint);
  set_(line1.direction, line1Direction);
  
  set_(line2.point, agentPosition);
  set_(line2.direction, blockingEdgeDir);
  
  const intersection = lineLineIntersection(line1, line2);
  
  if (!intersection) {
    return null;
  }
  
  const distanceToOriginalCornerSq = distance_sq(agentPosition, nextCorner);
  const distanceToIntersectionSq = distance_sq(agentPosition, intersection);
  
  if (distanceToIntersectionSq > distanceToOriginalCornerSq * 2.25) {
    return null;
  }
  
  return intersection;
}

const blockingEdgeDirection: Point2 = { x: 0, y: 0 };

/**
 * Merges multiple triangle corridors with an agent's existing corridor.
 * Converts triangle corridors to polygon corridors and handles proper joining.
 * 
 * @param navmesh - The navigation mesh
 * @param triangleCorridors - Array of triangle corridors to merge (in order)
 * @param agentCorridor - The agent's existing corridor
 * @param joinTriangle - The triangle where the new path rejoins the original corridor
 * @returns New merged corridor
 */
// Mirrors the C++ merge_corridors in path_patching.cpp
// triangleCorridors must be ordered so that the first starts at the join polygon
export function mergeCorridors(
  navmesh: Navmesh,
  triangleCorridorFirst: number[],
  triangleCorridorSecond: number[],
  agentCorridor: number[],
  joinTriangle: number
): number[] {
  const polygonCorridor: number[] = [];
  
  // First corridor must begin at the join polygon; convert triangles to polys in reverse order
  for (let i = triangleCorridorFirst.length - 1; i >= 0; i--) {
    const poly = navmesh.triangle_to_polygon[triangleCorridorFirst[i]];
    if (polygonCorridor.length === 0 || polygonCorridor[polygonCorridor.length - 1] !== poly) {
      polygonCorridor.push(poly);
    }
  }
  // Followed by the second corridor continuing back toward the agent
  for (let i = triangleCorridorSecond.length - 1; i >= 0; i--) {
    const poly = navmesh.triangle_to_polygon[triangleCorridorSecond[i]];
    if (polygonCorridor.length === 0 || polygonCorridor[polygonCorridor.length - 1] !== poly) {
      polygonCorridor.push(poly);
    }
  }
  
  let originalCorridorJoinIndex = -1;
  const joinPoly = navmesh.triangle_to_polygon[joinTriangle];
  for (let i = agentCorridor.length - 1; i >= 0; i--) {
    if (agentCorridor[i] === joinPoly) {
      originalCorridorJoinIndex = i;
      break;
    }
  }


  const restOfCorridor = originalCorridorJoinIndex === -1 ? [] : agentCorridor.slice(0, originalCorridorJoinIndex);
  const newCorridor = [...restOfCorridor, ...polygonCorridor];

  return newCorridor;
}

/**
 * Internal function to attempt geometric path patching using a failed raycast result.
 * This function implements the geometric path correction logic:
 * 1. Use the blocking edge from the failed raycast
 * 2. Calculate intersection R of line from last visible point to corner and line through agent parallel to blocking edge
 * 3. Validate R is reachable and on navmesh
 * 4. Update agent's corners if successful
 * 
 * @param navmesh - The navigation mesh
 * @param agent - The agent whose path to patch
 * @param raycastResult - The failed raycast result containing the blocking edge
 * @returns true if path was successfully patched, false if full repath is needed
 */
export function attemptPathPatchInternal(
  navmesh: Navmesh,
  agent: Agent,
  raycastResult: RaycastWithCorridorResult
): boolean {  
  if (raycastResult.hitV1_idx === -1) {
    return false;
  }

  const hitV1_idx = raycastResult.hitV1_idx;
  const hitV2_idx = raycastResult.hitV2_idx;
  const hitP1 = { x: navmesh.vertices[hitV1_idx * 2], y: navmesh.vertices[hitV1_idx * 2 + 1] };
  const hitP2 = { x: navmesh.vertices[hitV2_idx * 2], y: navmesh.vertices[hitV2_idx * 2 + 1] };
  
  // sceneState.addDebugCircle(copy(hitP1), ACINDIGO);
  // sceneState.addDebugCircle(copy(hitP2), ACINDIGO);

  if (raycastResult.hitTri_idx !== -1) {
    const blobIdx = navmesh.triangle_to_polygon[raycastResult.hitTri_idx];
    
    if (blobIdx >= navmesh.walkable_polygon_count) {
      
      const dist1Sq = distancePointToSegment(hitP1, agent.lastVisiblePointForNextCorner, agent.nextCorner);
      const dist2Sq = distancePointToSegment(hitP2, agent.lastVisiblePointForNextCorner, agent.nextCorner);

      const cornerPoint = dist1Sq < dist2Sq ? hitP1 : hitP2;
      const cornerVIdx = dist1Sq < dist2Sq ? hitV1_idx : hitV2_idx;

      const offsetPoint = offsetCorner(navmesh, cornerPoint, cornerVIdx, blobIdx, NavConst.CORNER_OFFSET);

      if (offsetPoint) {
        // sceneState.addDebugX(copy(offsetPoint), ACPINK);
        // sceneState.addDebugLine(copy(cornerPoint), copy(offsetPoint), ACPINK);
        
        const offsetTri = getTriangleFromPoint(navmesh, offsetPoint);
        
        if (offsetTri !== -1) {
          // sceneState.addDebugLine(copy(agent.coordinate), copy(offsetPoint), ACEMERALD);
          const raycastToOffset = raycastCorridor(navmesh, agent.coordinate, offsetPoint, agent.currentTri, offsetTri);
          
          if (raycastToOffset.hitV1_idx === -1 && raycastToOffset.corridor) {
            
            if (agent.numValidCorners === 2) {
              // sceneState.addDebugLine(copy(offsetPoint), copy(agent.nextCorner2), ACYELLOW);
              const raycastToNextCorner2 = raycastCorridor(navmesh, offsetPoint, agent.nextCorner2, offsetTri, agent.nextCorner2Tri);
              
              if (raycastToNextCorner2.hitV1_idx === -1 && raycastToNextCorner2.corridor) {
                // Great: we can go offset -> nextCorner2; set nextCorner to offset and keep nextCorner2
                set_(agent.nextCorner, offsetPoint);
                agent.nextCornerTri = offsetTri;
                // numValidCorners remains 2
                
                agent.corridor = mergeCorridors(
                  navmesh,
                  raycastToNextCorner2.corridor,
                  raycastToOffset.corridor,
                  agent.corridor,
                  agent.nextCorner2Tri
                );
                
                return true;
              } else {
                // fallback to trying nextCorner first
                // sceneState.addDebugLine(copy(offsetPoint), copy(agent.nextCorner), ACEMERALD);
                const raycastToNextCorner = raycastCorridor(navmesh, offsetPoint, agent.nextCorner, offsetTri, agent.nextCornerTri);
                if (raycastToNextCorner.hitV1_idx === -1 && raycastToNextCorner.corridor) {
                  // Update corners: insert offset as nextCorner, keep nextCorner2 as is
                  set_(agent.nextCorner, offsetPoint);
                  agent.nextCornerTri = offsetTri;
                  // numValidCorners remains 2

                  agent.corridor = mergeCorridors(
                    navmesh,
                    raycastToNextCorner.corridor,
                    raycastToOffset.corridor,
                    agent.corridor,
                    agent.nextCorner2Tri 
                  );
                  return true;
                }
                // give up miter approach
              }
            } else {
              // Only one corner known: must be able to go offset -> nextCorner
              // sceneState.addDebugLine(copy(offsetPoint), copy(agent.nextCorner), ACEMERALD);
              const raycastToNextCorner = raycastCorridor(navmesh, offsetPoint, agent.nextCorner, offsetTri, agent.nextCornerTri);
              
              if (raycastToNextCorner.hitV1_idx === -1 && raycastToNextCorner.corridor) {
                
                set_(agent.nextCorner2, agent.nextCorner);
                agent.nextCorner2Tri = agent.nextCornerTri;
                
                set_(agent.nextCorner, offsetPoint);
                agent.nextCornerTri = offsetTri;
                
                agent.numValidCorners = 2;
                
                agent.corridor = mergeCorridors(
                  navmesh,
                  raycastToNextCorner.corridor,
                  raycastToOffset.corridor,
                  agent.corridor,
                  agent.nextCorner2Tri
                );
                
                return true;
              }
            }
          }
        }
      }
    }
  }

  // sceneState.addDebugLine(copy(agent.lastVisiblePointForNextCorner), copy(agent.nextCorner), ACBLUE);
  
  set_(blockingEdgeDirection, hitP2);
  subtract_(blockingEdgeDirection, hitP1);
  normalize_(blockingEdgeDirection);
  
  const intersectionR = calculatePathPatchIntersection(
    agent.lastVisiblePointForNextCorner,
    agent.nextCorner,
    agent.coordinate,
    blockingEdgeDirection
  );
  
  if (!intersectionR) {
    return false;
  }

  // sceneState.addDebugX(copy(intersectionR), ACMAGENTA);
  // sceneState.addDebugLine(copy(agent.coordinate), intersectionR, ACBLACK);

  const rTriangle = getTriangleFromPoint(navmesh, intersectionR);
  
  if (rTriangle === -1) {
    return false;
  }

  const raycastToR = raycastCorridor(navmesh, agent.coordinate, intersectionR, agent.currentTri, rTriangle);
  
  if (raycastToR.hitV1_idx !== -1) {
    return false;
  }
  
  const raycastToNextCorner = raycastCorridor(navmesh, intersectionR, agent.nextCorner, rTriangle, agent.nextCornerTri);

  if (raycastToNextCorner.hitV1_idx !== -1) {
    return false;
  }
  
  set_(agent.nextCorner2, agent.nextCorner);
  agent.nextCorner2Tri = agent.nextCornerTri;
  
  set_(agent.nextCorner, intersectionR);
  agent.nextCornerTri = rTriangle;

  agent.numValidCorners = 2;
  
  if (raycastToR.corridor && raycastToNextCorner.corridor) {
    // Order matters: start with the segment that begins at the join poly (old nextCorner)
    agent.corridor = mergeCorridors(
      navmesh,
      raycastToNextCorner.corridor,
      raycastToR.corridor,
      agent.corridor,
      agent.nextCorner2Tri
    );
  }
  
  return true;
}
